File: contents/stacks_and_queues/code/csharp/Queue.cs
using System;
using System.Collections.Generic;

namespace my {
    /**
     * implementation using linked list
     * [value][next] -> [value][next] -> ... -> [value][next]
     * (top Node)      (intermediat Nodes)
     * left most Node represents top element of queue
     */
    public class Node {
        public Node next;
        public object data;
    }
    public class Queue {
        private Node head;
        private Node tail;
        public int size;

        public Queue() {
            head = new Node();
            tail = head;
            size = 0;
        }
        public void Enqueue(object value) {
            Node newNode = new Node();
            newNode.data = value;
            tail.next = newNode;
            tail = newNode;
            size++;
        }

        public void Dequeue() {
            if (size != 0) {
                head.next = head.next.next;
                size--;
            }
            else {
                Console.WriteLine("No element to remove.");
            }
        }

        public object Front() {
            return head.next.data;
        }

        public int Size() {
            return size;
        }

        public bool Empty() {
            return size == 0;
        }

    }

    class Test {
        static void Main(string[] args) {
            Console.WriteLine("Queue:");
            Queue intQueue = new Queue();

            intQueue.Enqueue(4);
            intQueue.Enqueue(5);
            intQueue.Enqueue(9);

            Console.Write("Size: ");
            Console.WriteLine(intQueue.Size());
            Console.Write("Front: ");
            Console.WriteLine(intQueue.Front());

            intQueue.Dequeue();

            Console.Write("Size: ");
            Console.WriteLine(intQueue.Size());
            Console.Write("Front: ");
            Console.WriteLine(intQueue.Front());

        }
    }
}

File: contents/stacks_and_queues/code/python/queue.py
#!/usr/bin/env python3

__author__ = "Michael Ciccotosto-Camp"

from typing import TypeVar, Generic


T = TypeVar("T")


class Queue(Generic[T]):
    def __init__(self) -> None:
        self.__list: list[T] = list()

    def dequeue(self) -> T:
        return self.__list.pop(0)

    def enqueue(self, element: T) -> int:
        self.__list.append(element)
        return len(self)

    def front(self) -> T:
        return self.__list[0]

    def __len__(self) -> int:
        return len(self.__list)

    def __str__(self) -> str:
        return str(self.__list)


def main() -> None:
    int_queue: Queue[int] = Queue()

    int_queue.enqueue(4)
    int_queue.enqueue(5)
    int_queue.enqueue(9)

    print(int_queue.dequeue())
    print(len(int_queue))
    print(int_queue.front())


if __name__ == "__main__":
    main()
